### ISSUE TITLE
@link tag inside a JSDoc type expression doesn't show parse error in TS

### ISSUE THREAD

**Issue by DanielSWolf** [microsoft/TypeScript#54650]
# Bug Report

### 🔎 Search Terms

tsdoc, `@throws`, `@link`, tags

### 🕗 Version & Regression Information

This changed between versions 4.9.5 and 5.0.4

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=5.0.4#code/MYGwhgzhAEBiD28CiAnF8XQKYA8AuWAdgCYyrqYDeAvgLABQokMAQmCuRtvkadJ1Tr0GAegBUYhtDHQAClhQAzDAFsYAIxABXdeoUA6KTKPQAAngAW6AO4xKpkAEtCAaziIB1aI8XRl8aAswAAdgoghDemkzSxs7B2c3Ng40DC8faHV2QJCwwgijEQZFLUJgPEd4QkztXQUACgBKaBogA)

### 💻 Code

```ts
class FooError extends Error {}
class BarError extends Error {}

/**
 * Performs blubber.
 * 
 * @throws {@link FooError} if foo happens.
 * @throws {@link BarError} if bar happens.
 */
function blubber() {}
```

### 🙁 Actual behavior

The `@link` tags after the `@throw` tags are parsed incorrectly, leading to this garbled tooltip:

![image](https://github.com/microsoft/TypeScript/assets/5963162/a205e811-e0b3-467d-a70f-51a62a039c87)

### 🙂 Expected behavior

The `@link` tags after the `@throw` tags are resolved correctly.

**Comment by fatcerberus** [comment]
Ah yes, performing blubber, the one feature *every* piece of software absolutely needs.

**Comment by sandersn** [comment]
The syntax for `throws` is `@throws {Type} documentation`, so the `{` starts a parse of a type expression, not a link tag.

All the other tags that are followed by a type have the same (or very similar) bug: `@param`, `@implements`, `@extends`, `@this`, etc.

In JS files with checkJS turned on, the parse error is shown. This is not true for TS files or plain JS files. At this point, maybe it makes sense to show parse errors from jsdoc, at least, in .ts files.

Unfortunately, that would require parsing jsdoc in .ts files. Which would prevent @jakebailey 's PR to disable jsdoc parsing in .ts files from working.

@DanielRosenwasser @gabritto do you have opinions? (I know we've discussed these issues in the past without reaching any consensus.)

**Comment by sandersn** [comment]
I put up a PR that skips `@link` at the start of jsdoc types too, since that would help this particular case, and seems like a reasonable mistake to make.

**Comment by sandersn** [comment]
I closed the PR -- it didn't work right and none of the 3 fixes were good enough.

**Comment by LuudJanssen** [comment]
@sandersn, I'm confused. The [TSDoc documentation for `@throws](https://tsdoc.org/pages/tags/throws/#docusaurus_skipToContent_fallback) shows almost the exact example @DanielRosenwasser posted. Has the spec since been updated? Should TypeScript now support this new TSDoc syntax?

**Comment by sandersn** [comment]
Which spec? Typescript's 'spec' is its code, and it has incomplete documentation at https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#typedef-callback-and-param (`@throws` isn't on there yet). jsdoc is documented at https://jsdoc.app/tags-throws. They both use the same syntax: `@throws {<type>} free-form description`. The type reference is optional but `{` after `throws` is interpreted as the start of a type reference, at lest in TS.

TSDoc is the different one here since its documentation indicates that its grammar is `@throws free-form description` -- to refer to a type you have to insert a reference to a symbol using `@link`.

I tried to support the TSDoc example in #55068 but failed. I think the TS syntax is reasonable, so maybe TSDoc is the one that should change.

**Comment by LuudJanssen** [comment]
Hi @sandersn thanks for coming back to me. Since TSDoc is also a Microsoft project, I assumed it was the official standard for writing comments in Typescript. Reading about the TSDoc project more, I realize it's a proposal to standardize the format, not it being the standard.

I get your issue with the syntax and I think I agree that it deviates unnecesarily from the JSDoc format.

@octogonz what do you think? The [`@throws` standard proposed by TSDoc](https://tsdoc.org/pages/tags/throws/) now isn't supported by TypeScript, and conflicts with JSDoc when using the `@link` tag enclosed by brackets at the start. Would it be a problem to have TSDoc support the syntax `@throws {<type>} <description>`?

**Comment by octogonz** [comment]
Sure, we can consider it.

However this conflicts somewhat with a couple design requirements of TSDoc:

- TSDoc "declaration references" (the target of `{@link ___}` or `{@inheritDoc ___}`) are currently designed to be relative to the package entry point and therefore not context sensitive, at least in API Extractor's implementation.  For example, if you move a `/** */` comment to a different file, the `{@link}` expressions don't need any changes.  They really refer to documentation objects, not TypeScript types, and this enables tools like API Documenter to resolve those links itself when processing `.api.json` files without needing any help from a TypeScript compiler.  

   By contrast, my understanding is that the `<type>` in `@throws {<type>}` is referring to a symbol in the current source file, which may be an aliased name, and probably won't resolve at all unless there is a corresponding `import`.  (And if the code doesn't use that type, would it be incorrectly flagged as an "unused" import?)  Or if `<type>` is just a freeform string with no validation, then how is the documentation website supposed to make a hyperlink?

- TSDoc aims to be an "extensible" syntax, which means that different tools can introduce their own custom tags while still agreeing about how a comment should be parsed.  For example, TSDoc defines `@throws` as a "block tag."  If a parser doesn't support `@throws` then it can safely discard all text after `@throws` until the next block tag, assuming that this entire section is part of the `@throws` block.  By contrast, Markdown is famously NOT extensible -- there is no way to introduce a custom extension (e.g. GitHub tables) without ruining other parser's ability to agree about basic questions like _"is this a real `{@link}` or is it quoted text?"_

   The `@throws {<type>} <description>` syntax seems like it might pose challenges for that.  For example, can you add generic parameters like `@throws {MyError<number>}`?  if so then `<number>` could be misinterpreted as an HTML tag.  Or is it freeform text between `{` `}`?  Then what if someone writes `@throws { \} }` -- how is that handled?
   
   
If you could point to a grammar/syntax spec for `@throws` that would be very helpful. Thanks!

**Comment by sandersn** [comment]
> my understanding is that the `<type>` in `@throws {<type>}` is referring to a symbol in the current source file, which may be an aliased name, and probably won't resolve at all unless there is a corresponding import. 

That is correct, it's a type reference evaluated the same way as any other one in that file. 
In a typescript file you can import a type that is only used in jsdoc, and it should mark it as used, not unused.

> The `@throws {<type>} <description>` syntax seems like it might pose challenges for that. 

Do TSDoc's custom tags end on both `@` and HTML tags? Typescript's tags end only on `@` as far as I remember.

Typescript parses type references in nearly every tag, so this would be a widespread mismatch, if it is one.

> If you could point to a grammar/syntax spec for @throws that would be very helpful. Thanks!

Here are the entry points for throws specifically, and jsdoc type references generally:
- [code for parseThrowsTag](https://github.com/microsoft/TypeScript/blob/2de69b09c5a67ab1e73ca952151521b5ca876dd9/src/compiler/parser.ts#L9413)
- [code for parseJSDocTypeExpression](https://github.com/microsoft/TypeScript/blob/2de69b09c5a67ab1e73ca952151521b5ca876dd9/src/compiler/parser.ts#L8729)
